### PR TITLE
docs: describe the purpose of the _dev dirs

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
It describes the different type of `_dev` dirs that are supported when building and testing an integration package.

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.

### CODE REVIEW

**Review by cachedout — APPROVED** [review]

File: docs/developer_workflow_design_build_test_integration.md [hunk 1]
```diff
@@ -1,5 +1,12 @@
 # Developer workflow: build and test the integration
 
+## Directory layout for tests
```
Comment (mtojek) [inline docs/developer_workflow_design_build_test_integration.md:3]: Let's rename it to: Development Extension ("_dev" directories)

File: docs/developer_workflow_design_build_test_integration.md [hunk 2]
```diff
@@ -1,5 +1,12 @@
...
+
+The `_dev` folder is an extension to [the standard package spec](https://github.com/elastic/package-spec), which has been created for development purposes. Currently, there are two _dev directories:
```
Comment (mtojek) [inline docs/developer_workflow_design_build_test_integration.md:5]: > Currently, there are two _dev directories:

Maybe rephrase it: The `_dev` directories can be defined on the following levels:

File: docs/developer_workflow_design_build_test_integration.md [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
+
+1. **Package** level (https://github.com/elastic/package-spec/tree/master/versions/1/_dev) - it contains resources to "build" docs content and "deploy" services for system tests.
+1. **Data-stream** level (https://github.com/elastic/package-spec/tree/master/versions/1/data_stream/_dev) - it contains "test" definitions for pipeline and system tests.
```
Comment (mtojek) [inline docs/developer_workflow_design_build_test_integration.md:8]: definitions for, i.e. pipeline and system tests.

Ok, let me explain:
We have also asset and static tests. They don't require config files, but configs can be used to mark them as optional.

File: docs/developer_workflow_design_build_test_integration.md [hunk 4]
```diff
@@ -1,5 +1,12 @@
 # Developer workflow: build and test the integration
 
+## Directory layout for tests
+
+The `_dev` folder is an extension to [the standard package spec](https://github.com/elastic/package-spec), which has been created for development purposes. Currently, there are two _dev directories:
+
+1. **Package** level (https://github.com/elastic/package-spec/tree/master/versions/1/_dev) - it contains resources to "build" docs content and "deploy" services for system tests.
```
Comment (mtojek) [inline docs/developer_workflow_design_build_test_integration.md:7]: Would you mind replacing the `_dev` urls with spec hyperlinks: https://github.com/elastic/package-spec/blob/master/versions/1/data_stream/_dev/spec.yml ?

This way the user will jump the exact place where these resources are defined. It may be less confusing.

**Review by mtojek — COMMENTED** [review]
Thanks for filling the gap, @mdelapenya , appreciated.

I would be for moving "Development Extension" to the Definitions.md section. 

This file is more oriented on the process/next steps, commands etc. than describing parts of the package format.

**Comment by mtojek** [comment]
I'm adding @ycombinator as he used to be responsible for docs proofreading :)

File: docs/developer_workflow_design_build_test_integration.md (hunk 2 above)
Comment (mdelapenya) [inline docs/developer_workflow_design_build_test_integration.md:5]:
```suggestion
The `_dev` directory is an extension to [the standard package spec](https://github.com/elastic/package-spec), which has been created for development purposes. It can be defined on the following levels:
```

File: docs/developer_workflow_design_build_test_integration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply docs/developer_workflow_design_build_test_integration.md:8]: Sorry I wasn't clear enough. I suggested to replace the URLs with hyperlinks:

[spec.yml](https://github.com/elastic/package-spec/tree/master/versions/1/_dev/spec.yml) - this form would be much shorter.

File: docs/developer_workflow_design_build_test_integration.md (hunk 1 above)
Comment (mdelapenya) [inline docs/developer_workflow_design_build_test_integration.md:3]:
```suggestion
## Development Extensions: '_dev' directories
```

File: docs/developer_workflow_design_build_test_integration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mtojek) [reply docs/developer_workflow_design_build_test_integration.md:8]: **Package** level ([spec.yml](https://github.com/elastic/package-spec/tree/master/versions/1/_dev/spec.yml)):  the directory contains ...

File: docs/developer_workflow_design_build_test_integration.md [hunk 5]
```diff
@@ -1,5 +1,14 @@
 # Developer workflow: build and test the integration
 
+## Development Extensions: '_dev' directories
```
Comment (mtojek) [inline docs/developer_workflow_design_build_test_integration.md:3]: As I suggested [here](https://github.com/elastic/integrations/pull/954#pullrequestreview-646772651) - it would be more consistent to put it under `definitions.md`.

Comment (mdelapenya) [reply docs/developer_workflow_design_build_test_integration.md:3]: Ahh sorry, I missed that comment :/ I'll push a change with your suggestion

**Review by mtojek — APPROVED** [review]
Thank you adjusting the PR, LGTM. Let's wait for the final approval from @ycombinator .

File: docs/definitions.md [hunk 6]
```diff
@@ -33,6 +33,15 @@ The data stream consists of:
 * Zero or more ingest pipelines
 * An Elastic Agent policy template
 
+## Development Extensions: '_dev' directories
+
+The `_dev` directory is an extension to [the standard package spec](https://github.com/elastic/package-spec), which has been created for development purposes. It can be defined on the following levels:
```
Comment (ycombinator) [inline docs/definitions.md:38]: This is not quite accurate. The `_dev` folders are actually covered by the package spec itself. See:
- https://github.com/elastic/package-spec/blob/7bf0ce9f5e865e01581a8c2dfa2adefbae0c2bf2/versions/1/spec.yml#L47-L52
- https://github.com/elastic/package-spec/blob/7bf0ce9f5e865e01581a8c2dfa2adefbae0c2bf2/versions/1/data_stream/spec.yml#L78-L83

Comment (ycombinator) [reply docs/definitions.md:38]: So I wouldn't call them extensions to the package spec. Also, saying "standard" package spec implies there is a... non-standard package spec?

Also, the package spec uses the terminology, "folder containing development resources". For consistency, lets use the same terminology here.

File: docs/definitions.md [hunk 7]
```diff
@@ -33,6 +33,15 @@ The data stream consists of:
...
+
+1. **Package** level ([spec.yml](https://github.com/elastic/package-spec/tree/master/versions/1/_dev/spec.yml)): the directory contains resources to "build" docs content and "deploy" services for system tests.
+1. **Data-stream** level ([spec.yml](https://github.com/elastic/package-spec/tree/master/versions/1/data_stream/_dev/spec.yml)): the directory contains "test" definitions for, i.e. pipeline and system tests.
+
+>The integrations have also asset and static tests. They don't require config files, but configs can be used to mark them as optional.
```
Comment (ycombinator) [inline docs/definitions.md:43]: I'm not sure why this is formatted as a quote. Is this sentence being quoted from somewhere else?

File: docs/definitions.md (hunk 6 above)
Comment (mdelapenya) [reply docs/definitions.md:38]: Is it possible to describe here what `development resources` means? Our intention os to clarify how to extend the extension from testing standpoint. If you can add a _suggestion_ in the file itself, that would be superb 😊

File: docs/definitions.md (hunk 7 above)
Comment (mdelapenya) [inline docs/definitions.md:43]:
```suggestion
The integrations have also asset and static tests. They don't require config files, but configs can be used to mark them as optional.
```

File: docs/definitions.md (hunk 6 above)
Comment (ycombinator) [reply docs/definitions.md:38]: Yeah, the use of "development resources" in the package spec is deliberately generic. It's intended to cover any types of files/folders needed only at development time. This includes resources needed for testing but also includes any templates that might be used for generating documentation.

Focussing on testing:
- the package-level `_dev` folder contains files needed to setup the testing environment for that package. This environment setup is specified via folders/files in the  `_dev/deploy` folder. For example, the `apache` package [specifies](https://github.com/elastic/integrations/tree/master/packages/apache/_dev/deploy) how to spin up an Apache Docker container for testing.
- the data stream-level `_dev` folder contains test configuration files for various types of tests. For example, see the [`_dev/test` folder](https://github.com/elastic/integrations/tree/master/packages/apache/data_stream/error/_dev/test) under the `apache/error` data stream.

Comment (mdelapenya) [reply docs/definitions.md:38]: This is very well explained, thanks! I'll add it to the docs now

File: docs/definitions.md [hunk 8]
```diff
@@ -33,6 +33,15 @@ The data stream consists of:
 * Zero or more ingest pipelines
 * An Elastic Agent policy template
 
+## Development Extensions: '_dev' directories
+
+The `_dev` directory is an extension to [the standard package spec](https://github.com/elastic/package-spec), which has been created for development purposes. It can be defined on the following levels:
+
+1. **Package** level ([spec.yml](https://github.com/elastic/package-spec/tree/master/versions/1/_dev/spec.yml)): the directory contains resources to "build" docs content and "deploy" services for system tests.
+1. **Data-stream** level ([spec.yml](https://github.com/elastic/package-spec/tree/master/versions/1/data_stream/_dev/spec.yml)): the directory contains "test" definitions for, i.e. pipeline and system tests.
```
Comment (mdelapenya) [inline docs/definitions.md:41]:
```suggestion
1. the package-level `_dev` folder contains files needed to setup the testing environment for that package. This environment setup is specified via folders/files in the `_dev/deploy` folder. For example, the `apache` package [specifies](https://github.com/elastic/integrations/tree/master/packages/apache/_dev/deploy) how to spin up an Apache Docker container for testing.
1. the data stream-level `_dev` folder contains test configuration files for various types of tests. For example, see the [`_dev/test` folder](https://github.com/elastic/integrations/tree/master/packages/apache/data_stream/error/_dev/test) under the `apache/error` data stream.
```

File: docs/definitions.md (hunk 6 above)
Comment (mdelapenya) [inline docs/definitions.md:38]:
```suggestion
The `_dev` directory is an extension to [the package spec](https://github.com/elastic/package-spec), containing development resources. These development resources cover any types of files/folders needed only at development time. This includes resources needed for testing but also includes any templates that might be used for generating documentation. In the future it could include other files/folders needed just at development time. It can be defined on the following levels:
```

Comment (ycombinator) [reply docs/definitions.md:38]: Again, let's please not call it an "extension" to the package spec. That implies that the `_dev` directory is not part of the package spec itself, which is not accurate.

Comment (mdelapenya) [reply docs/definitions.md:38]: I'll remove it. This sentence was provided by the team and I just pasted it in the editor 😃

Comment (mtojek) [reply docs/definitions.md:38]: Sorry, that was me. It's true that there is no official extensions to the spec, but I used to call it like this informally. This is because we don't copy `_dev` resources to package-storage (it contains only the package files used by either Agent, Kibana or EPR).

I'm fine with renaming it, although I like the term "extension" :)

Comment (ycombinator) [reply docs/definitions.md:38]: I understand the informal use but using inconsistent terminology or concepts in different places (here vs. package spec) is going to cause confusion for readers who encounter both. Let's reserve "extension" for if/when we actually need an extension 😄. 

I like the latest suggestion, where we simply say:

> the `_dev` directory is part of the package spec, containing development resources.

And then we explain what those resources can be.

It explains the purpose of the folder without introducing new and potentially confusing terminology. 👍

File: docs/definitions.md [hunk 9]
```diff
@@ -33,6 +33,15 @@ The data stream consists of:
...
 * An Elastic Agent policy template
 
+## Development Extensions: '_dev' directories
+
+The `_dev` directory is an extension to [the standard package spec](https://github.com/elastic/package-spec), which has been created for development purposes. It can be defined on the following levels:
+
+1. the package-level `_dev` folder contains files needed to setup the testing environment for that package. This environment setup is specified via folders/files in the `_dev/deploy` folder. For example, the `apache` package [specifies](https://github.com/elastic/integrations/tree/master/packages/apache/_dev/deploy) how to spin up an Apache Docker container for testing.
+1. the data stream-level `_dev` folder contains test configuration files for various types of tests. For example, see the [`_dev/test` folder](https://github.com/elastic/integrations/tree/master/packages/apache/data_stream/error/_dev/test) under the `apache/error` data stream.
+
+The integrations have also asset and static tests. They don't require config files, but configs can be used to mark them as optional.
```
Comment (ycombinator) [inline docs/definitions.md:43]:
```suggestion
The integrations have also [asset](https://github.com/elastic/elastic-package/blob/master/docs/howto/asset_testing.md) and [static](https://github.com/elastic/elastic-package/blob/master/docs/howto/static_testing.md) tests. They don't require config files, but configs can be used to mark them as optional.
```

**Review by ycombinator — APPROVED** [review]
LGTM once pending suggestions are accepted.

**Comment by mdelapenya** [comment]
If you agree, I'm going to merge this one. Thanks for the review!!

**Review by ycombinator — APPROVED** [review]
LGTM.
